### ISSUE TITLE
Folders failing to open on remote ssh after they have been closed once

### ISSUE THREAD

**Issue by tanja84dk** [microsoft/vscode-remote-release#7438]
- VSCode Version: 1.72.2
- Local OS Version: Windows 10 Pro
- Remote OS Version: Ubuntu 22.04 ( Behind NAT because it is a docker container on my home server with a portforwarding port  2287 is forwarded to the servers port 22 )
- Remote SSH: v0.90.1 from marketplace (https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-ssh)
- Logs: [first login.log](https://github.com/microsoft/vscode-remote-release/files/9896729/first.login.log) [Second login when opening the folder.log](https://github.com/microsoft/vscode-remote-release/files/9896730/Second.login.when.opening.the.folder.log) [ssh config on windows.txt](https://github.com/microsoft/vscode-remote-release/files/9896732/ssh.config.on.windows.txt)

I did try to first deactivate all extentions locally extept the Remote SSH ( No change ), then also uninstall all extensions and reinstall Remote SSH ( No Change ), and also literally delete the .vscode folder in %userprofile% and also delete the usersettings file located %appdata%\Code\User\settings.json to literally forcing to reset everything if it were because of a settings issue but also no change.

My only hunch is that vscode internally somehow cache/index past connections and only does it based on ip/hostname and disregard the port there is used which could make issues in situations like here where the connections are natted on the receiving end

And as a side note most of my projects are c++ based and that is where I at least have faced it and I did also try to the home server it self and do the programming there but that gave the same result as in second time I opened the folder on remote then it failed.

Steps to Reproduce:

1. Connect to the remote SSH ( Works fine every time )
2. Opening a project folder to work in ( Works the first time and every time I try to reopen the folder again then then it times out in the ssh connection even that it connects fine to the Remote SSH as long I dont open the folder )

Does this issue occur when you try this locally?: No because its at connection based
Does this issue occur when you try this locally and all extensions are disabled?: Yes/No

**Comment by tanja84dk** [comment]
Just an additional information.
I have just tried to completely uninstall Visual Studio Code and make sure all the folders were deleted including %appdata%\Code and %userprofile%/.vscode and then reinstall it again fresh and unfortunally even with only Remote - SSH and Remote - SSH: Editing Configuration Files installed it still fails the same way ( Successfully opens the folder once and then times out from the 2 time I open the folder )

**Comment by strazto** [comment]
On `v0.90.1` I've been getting something similar -

Once I open / close a workspace on the remote, I can never open that workspace again on the remote.

If I want to open it again, my best workaround is currently to rename the folder from `ssh` command line (Either direct `ssh` terminal connection, or from the integrated terminal in a vscode session that isn't broken), eg

```bash
# imagine ~/my-workspace was previously working
code ~/my-workspace
# Run command: Developer: Reload Window
# No connection

# the following times out
code ~/my-workspace 

mv ~/my-workspace ~/my-workspace-2

# The following will work, but only once
code ~/my-workspace-2
```

**Comment by strazto** [comment]
When running windows, deleting `%appdata%/Code` allows reconnection to the workspace.

**Comment by tanja84dk** [comment]
Thank you very much for your input and yeah its weird that t happens and I'm also on 0.90.1.
And thank you very much for letting me know about the workarounds you have found and I think i'll try downgrade and clean the broken state first and if it still gives issues then use the renaming of folder workaround

**Comment by tanja84dk** [comment]
Uodate: unfortunally after vsvode closed for a couple of hours and wanted to access the workfolder then it happened again with the same error and that was on 0.88.0 so yeah I'm starting to give up on it

**Comment by eleanorjboyd** [comment]
Hello! What happens if you use the command "kill server on remote"? Additionally can you send the remote-ssh logs and the logs from the developer tools? This could be useful in providing an error message that gives direction. Thanks!

**Comment by tanja84dk** [comment]
@eleanorjboyd I infortunally dont know of how to get those logs but If you look in the first post the logs are attached from within Visual Studio Code from the Remote SSH log output and also the ssh client host config from my windows 10

And about the "Kill server on remote" Tbh I have no idea about how etc because I dont know where or what it is, but I do also have to say it wouldnt help because it does connect to the remote ssh as long you dont open a workspace you prior has opened or connect without connecting to a workspace/folder and it doesnt help removing and reinstalling the vscode-server on the remote ssg. The only thing there is temporarily fixing it is on the pc i'm using I then has to delete some data within %appdata%/Code and ecause I dont know which then the only fix is to delete the full %ppdata%/Code for then to be able to connect to folders I prior has connected to with remote ssh again ( Because it fails by saying its timing out the second time a folder is accessed on 0.90.1 ) ( just as the way @strazto outlined in his first respons )

Also I have noticed why its failing after having downgraded and then restarts the program because I have just noticed Visual Studio Code literally automatically upgrades the addons for some reason to the version there is the newest and is failing

Su to be real the way I'm working around it at least today is that I have to delete all my configs ( because of its stored in %appdata%/Code ) every time I have to work in vs code wich unfortunally then also deletes all my settings about line endings, compiler settings, theme, default terminal etc so literally a pain

**Comment by strazto** [comment]
@eleanorjboyd https://github.com/microsoft/vscode-remote-release/issues/7438#issuecomment-1297364174
> Hello! What happens if you use the command "kill server on remote"?

I tried this while debugging, by killing the server from other (working) vscode sessions (eg, a session that doesn't have a workspace, or a workspace that isn't yet broken).

I also tried some more forceful methods, along the lines of shell stuff over a plain ssh session:

```bash
# searching for vscode server running on the remote and kill it
ps -aex | search_for_vscode_somehow
kill vscode_server_pids

# some kind of socket / state cleanup on the remote
rm some_vscode_sockets 
```

I didn't document what I actually did for the above, but nothing had any real impact to speak of.

**Comment by strazto** [comment]
@tanja84dk 
> Su to be real the way I'm working around it at least today is that I have to delete all my configs ( because of its stored in `%appdata%/Code` ) every time I have to work in vs code wich unfortunally then also deletes all my settings about line endings, compiler settings, theme, default terminal etc so literally a pain

Quick tip for this, if you use settings sync you can get that stuff back a lot easier, but also, I suspect the broken state is actually in one of the "cache" type files

**Comment by strazto** [comment]
@eleanorjboyd 
> Additionally can you send the remote-ssh logs and the logs from the developer tools? 


#### Remote - SSH log (from output panel) 

Doesn't look too useful

<details>
<summary>Remote - SSH log</summary>

```log
[11:16:52.365] Log Level: 2
[11:16:52.367] remote-ssh@0.88.0
[11:16:52.368] darwin arm64
[11:16:52.380] SSH Resolver called for "ssh-remote+myremote-external", attempt 1
[11:16:52.380] "remote.SSH.useLocalServer": false
[11:16:52.381] "remote.SSH.showLoginTerminal": false
[11:16:52.381] "remote.SSH.remotePlatform": {"windows.portal":"windows","me":"windows","server.portal":"linux","myremote":"linux","myremote-external":"linux"}
[11:16:52.381] "remote.SSH.path": undefined
[11:16:52.381] "remote.SSH.configFile": undefined
[11:16:52.381] "remote.SSH.useFlock": true
[11:16:52.381] "remote.SSH.lockfilesInTmp": false
[11:16:52.381] "remote.SSH.localServerDownload": off
[11:16:52.381] "remote.SSH.remoteServerListenOnSocket": false
[11:16:52.381] "remote.SSH.showLoginTerminal": false
[11:16:52.381] "remote.SSH.defaultExtensions": []
[11:16:52.381] "remote.SSH.loglevel": 2
[11:16:52.381] "remote.SSH.enableDynamicForwarding": true
[11:16:52.382] "remote.SSH.enableRemoteCommand": false
[11:16:52.382] "remote.SSH.serverPickPortsFromRange": {}
[11:16:52.382] "remote.SSH.serverInstallPath": {}
[11:16:52.390] SSH Resolver called for host: myremote-external
[11:16:52.390] Setting up SSH remote "myremote-external"
[11:16:52.392] Using commit id "d045a5eda657f4d7b676dedbfa7aab8207f8a075" and quality "stable" for server
[11:16:52.393] Install and start server if needed
[11:16:52.398] PATH: /Users/matty/.poetry/bin:/opt/local/bin:/opt/local/sbin:/usr/local/Cellar/vim/8.2.0200/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/Users/matty/.local/bin:/Users/matty/.poetry/bin:/Users/matty/.local/bin:/Users/matty/perl5/bin:/Users/matty/.nvm/versions/node/v18.10.0/bin:/Users/matty/gems/bin:/usr/local/go/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin:/Users/matty/go/bin:/opt/homebrew/opt/fzf/bin:/bin
[11:16:52.398] Checking ssh with "ssh -V"
[11:16:52.404] > OpenSSH_8.6p1, LibreSSL 3.3.6

[11:16:52.407] Running script with connection command: ssh -T -D 56521 -o ConnectTimeout=40 'myremote-external' bash
[11:17:34.412] Resolver error: Error: Connecting with SSH timed out
	at g.Timeout (/Users/matty/.vscode/extensions/ms-vscode-remote.remote-ssh-0.88.0/out/extension.js:1:583947)
	at Timeout._onTimeout (/Users/matty/.vscode/extensions/ms-vscode-remote.remote-ssh-0.88.0/out/extension.js:1:678419)
	at listOnTimeout (node:internal/timers:559:17)
	at process.processTimers (node:internal/timers:502:7)
[11:17:34.435] ------




[11:17:34.450] ------




[11:17:34.450] SSH Resolver called for "ssh-remote+myremote-external", attempt 2, (Reconnection)
[11:17:34.454] SSH Resolver called for host: myremote-external
[11:17:34.454] Setting up SSH remote "myremote-external"
[11:17:34.454] Using commit id "d045a5eda657f4d7b676dedbfa7aab8207f8a075" and quality "stable" for server
[11:17:34.457] Install and start server if needed
[11:17:34.462] Running script with connection command: ssh -T -D 56539 -o ConnectTimeout=40 'myremote-external' bash
[11:18:16.464] Resolver error: Error: Connecting with SSH timed out
	at g.Timeout (/Users/matty/.vscode/extensions/ms-vscode-remote.remote-ssh-0.88.0/out/extension.js:1:583947)
	at Timeout._onTimeout (/Users/matty/.vscode/extensions/ms-vscode-remote.remote-ssh-0.88.0/out/extension.js:1:678419)
	at listOnTimeout (node:internal/timers:559:17)
	at process.processTimers (node:internal/timers:502:7)
[11:18:16.494] ------




[11:18:16.508] ------




[11:18:16.508] SSH Resolver called for "ssh-remote+myremote-external", attempt 3, (Reconnection)
[11:18:16.511] SSH Resolver called for host: myremote-external
[11:18:16.511] Setting up SSH remote "myremote-external"
[11:18:16.512] Using commit id "d045a5eda657f4d7b676dedbfa7aab8207f8a075" and quality "stable" for server
[11:18:16.513] Install and start server if needed
[11:18:16.517] Running script with connection command: ssh -T -D 56542 -o ConnectTimeout=40 'myremote-external' bash
[11:18:58.519] Resolver error: Error: Connecting with SSH timed out
	at g.Timeout (/Users/matty/.vscode/extensions/ms-vscode-remote.remote-ssh-0.88.0/out/extension.js:1:583947)
	at Timeout._onTimeout (/Users/matty/.vscode/extensions/ms-vscode-remote.remote-ssh-0.88.0/out/extension.js:1:678419)
	at listOnTimeout (node:internal/timers:559:17)
	at process.processTimers (node:internal/timers:502:7)
[11:18:58.527] ------




[11:18:58.547] ------




[11:18:58.547] SSH Resolver called for "ssh-remote+myremote-external", attempt 4, (Reconnection)
[11:18:58.549] SSH Resolver called for host: myremote-external
[11:18:58.549] Setting up SSH remote "myremote-external"
[11:18:58.550] Using commit id "d045a5eda657f4d7b676dedbfa7aab8207f8a075" and quality "stable" for server
[11:18:58.551] Install and start server if needed
[11:18:58.555] Running script with connection command: ssh -T -D 56545 -o ConnectTimeout=40 'myremote-external' bash
[11:19:40.558] Resolver error: Error: Connecting with SSH timed out
	at g.Timeout (/Users/matty/.vscode/extensions/ms-vscode-remote.remote-ssh-0.88.0/out/extension.js:1:583947)
	at Timeout._onTimeout (/Users/matty/.vscode/extensions/ms-vscode-remote.remote-ssh-0.88.0/out/extension.js:1:678419)
	at listOnTimeout (node:internal/timers:559:17)
	at process.processTimers (node:internal/timers:502:7)
[11:19:40.579] ------




[11:19:40.603] ------




[11:19:40.603] SSH Resolver called for "ssh-remote+myremote-external", attempt 5, (Reconnection)
[11:19:40.605] SSH Resolver called for host: myremote-external
[11:19:40.605] Setting up SSH remote "myremote-external"
[11:19:40.606] Using commit id "d045a5eda657f4d7b676dedbfa7aab8207f8a075" and quality "stable" for server
[11:19:40.607] Install and start server if needed
[11:19:40.610] Running script with connection command: ssh -T -D 56557 -o ConnectTimeout=40 'myremote-external' bash
```

</details>

**Comment by strazto** [comment]
Update:

The following doesn't clear the broken state:

```
# Local MacOS

cd "$HOME/Library/Application Support/Code"

# Delete anything
rm -r Cache*

# Relaunch vscode, still timeout
```

**Comment by strazto** [comment]
Found a less destructive workaround for @tanja84dk :

```bash
# osx:
rm -r "$HOME/Library/Application Support/Code/User/workspaceStorage/"

```

For windows, I forget how variable substitution works, but delete `%appdata%/Code/User/workspaceStorage` 

I'm pretty sure we lose some state when we do this, so it IS possible to lose work, etc, but I'm pretty sure this storage tends to be temporary state like unsaved changes in editor etc. 

How negative of an impact this is depends on whether you meant to terminate your session when you did (For example, I'd be peeved if I just wanted to reload my window because one of my extensions was playing up, and I had unsaved changes open in the editor)

**Comment by RBedbur** [comment]
I had the same problem with only one folder on my remote server. I was able to connect with remote ssh an browse the folder. I could browse to the specific folder, but opening failed. Opening another folder even parent folder or subfolder worked. In the end I solved it this way (like comments above):

- In my remote ssh output in vs code was a line 'using commit id "8f..." and quality "stable for server'. In my app-data under c:\user\<username>\AppData\Roaming\Code\CachedData\ I deleted the folder with the commit id.
- deleted all folders under c:\user\<username>\AppData\Roaming\Code\User\workspaceStorage
- downgraded to remote ssh version 0.88 from 0.90.1

**Comment by strazto** [comment]
Yep, `0.88` (after applying https://github.com/microsoft/vscode-remote-release/issues/7438#issuecomment-1297865154 ) I've not had the problem

**Comment by rahedges** [comment]
I have been having this problem as well and have tried many things.  Deleting ~/.vscode-server on the remote did not, renaming the remote folder worked but was inconvenient.   Deleting the workspacesStorage as mentioned above https://github.com/microsoft/vscode-remote-release/issues/7438#issuecomment-1297865154  worked for me too.

**Comment by eleanorjboyd** [comment]
Hello! This is a duplicate of the discussion here: https://github.com/microsoft/vscode-remote-release/issues/7324 Which we are working on fixing. I am closing this issue in favor of that one and we will post updates there. Thanks!

**Comment by xgdgsc** [comment]
Just a note I had to close all windows except one opened on one server and delete the workspaceStorage folder. Otherwise it would still loop retrying/ waiting for other windows connect.
